README.md
# ECSx.LiveDashboard

[![Hex Version](https://img.shields.io/hexpm/v/ecsx_live_dashboard.svg)](https://hex.pm/packages/ecsx_live_dashboard)
[![License](https://img.shields.io/hexpm/l/ecsx_live_dashboard.svg)](https://github.com/APB9785/ecsx_live_dashboard/blob/master/LICENSE)

![ECSx LiveDashboard](assets/images/screenshot.png)

Adds an [ECSx](https://github.com/APB9785/ECSx) page to Phoenix LiveDashboard

## Features

* Monitor Systems load compared to your app's maximum capability at the current tick rate
* Compare the load of each individual System to identify where optimizations would be most beneficial
* Observe the quantity of each Component type, to ensure Components are being created and/or deleted properly
* Inspect current Component values to check data integrity

## Requirements

Your ECSx application must use Phoenix and have Phoenix LiveDashboard 0.8+ installed

## Installation

The package can be installed by adding `ecsx_live_dashboard` to your list of dependencies in `mix.exs`

```elixir
def deps do
  [
    ...
    {:ecsx_live_dashboard, "~> 0.1"},
    ...
  ]
end
```

Then modify the `live_dashboard` call in your `router.ex` to include ECSx LiveDashboard as an additional page:

```elixir
live_dashboard "/dashboard",
  metrics: MyAppWeb.Telemetry,
  additional_pages: [
    ecsx: ECSx.LiveDashboard.Page
  ]
```

## License

Copyright (C) 2023  Andrew P Berrien

This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the [GNU General Public License](https://www.gnu.org/licenses/gpl.html) for more details.


CHANGELOG.md
# Changelog

## v0.1.0 (2023-07-12)

  * Overall systems load
  * Systems page with pie chart
  * Component types page with quantities
  * List of all components of a certain type

test/ecsx_dashboard_test.exs
defmodule ECSx.LiveDashboardTest do
  use ExUnit.Case
end


mix.exs
defmodule ECSx.LiveDashboard.MixProject do
  use Mix.Project

  @gh_url "https://github.com/APB9785/ecsx_live_dashboard"
  @version "0.1.0"

  def project do
    [
      app: :ecsx_live_dashboard,
      version: @version,
      elixir: "~> 1.13",
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      package: package(),
      name: "ECSx Live Dashboard",
      docs: docs(),
      description: "Adds an ECSx page to Phoenix Live Dashboard"
    ]
  end

  def application do
    [
      mod: {ECSx.LiveDashboard, []},
      extra_applications: [:logger]
    ]
  end

  defp docs do
    [
      main: "readme",
      source_ref: "v#{@version}",
      extras: ["README.md"],
      logo: nil,
      source_url: @gh_url
    ]
  end

  defp package do
    [
      maintainers: ["Andrew P Berrien"],
      licenses: ["GPL-3.0"],
      links: %{
        "Changelog" => "#{@gh_url}/blob/master/CHANGELOG.md",
        "GitHub" => @gh_url
      }
    ]
  end

  defp deps do
    [
      {:ecsx, "~> 0.4"},
      {:phoenix_live_dashboard, "~> 0.8"},
      {:contex, "~> 0.4"}
    ]
  end
end


lib/ecsx/live_dashboard/store.ex
defmodule ECSx.LiveDashboard.Store do
  use GenServer

  @system_buffer_size 10

  def start_link(_) do
    GenServer.start_link(__MODULE__, [], name: __MODULE__)
  end

  def init(_) do
    state = %{
      systems: %{},
      components: %{},
      client_events: %{min: :infinity, max: 0, store: []}
    }

    IO.inspect("dashboard store started!")

    {:ok, state}
  end

  def handle_call(:get_systems, _from, state) do
    {:reply, state.systems, state}
  end

  def handle_call(:get_components, _from, state) do
    {:reply, state.components, state}
  end

  def handle_cast({:system_run, name, duration_us}, state) do
    {:noreply, %{state | systems: update_systems(state.systems, name, duration_us)}}
  end

  def handle_cast({:component_action, name, action, second}, state) do
    new_components =
      Map.update(state.components, name, %{action => %{second => 1}}, fn action_map ->
        Map.update(action_map, action, %{second => 1}, fn seconds_map ->
          Map.update(seconds_map, second, 1, &(&1 + 1))
        end)
      end)

    {:noreply, %{state | components: new_components}}
  end

  def handle_cast({:client_events, _count}, state) do
    {:noreply, state}
  end

  defp update_systems(systems, name, duration) do
    Map.update(
      systems,
      name,
      %{min: duration, max: duration, buffer: List.duplicate(duration, @system_buffer_size)},
      &add_system_tick(&1, duration)
    )
  end

  defp add_system_tick(%{min: min, max: max, buffer: buffer}, duration) do
    new_buffer = Enum.take([duration | buffer], @system_buffer_size)
    new_min = min(min, duration)
    new_max = max(max, duration)

    %{min: new_min, max: new_max, buffer: new_buffer}
  end

  # API

  def system_run(%{name: name, duration_us: duration_us}),
    do: GenServer.cast(__MODULE__, {:system_run, name, duration_us})

  def component_action(%{name: name, action: action, second: second}),
    do: GenServer.cast(__MODULE__, {:component_action, name, action, second})

  def client_events(%{count: count}),
    do: GenServer.cast(__MODULE__, {:client_events, count})

  def get_systems do
    systems = GenServer.call(__MODULE__, :get_systems)

    for {name, %{min: min, max: max, buffer: buffer}} <- systems do
      %{
        name: name,
        min: min,
        max: max,
        avg: Enum.sum(buffer) / @system_buffer_size
      }
    end
  end
end


lib/ecsx/live_dashboard/component.ex
defmodule ECSx.LiveDashboard.Component do
  use Phoenix.LiveComponent
  use Phoenix.HTML

  alias ECSx.LiveDashboard.Store

  def update(_, socket) do
    systems = Store.get_systems()
    tick_interval = div(1_000_000, ECSx.tick_rate())

    total_used =
      systems
      |> Enum.map(& &1.avg)
      |> Enum.sum()

    idle = %{name: "Idle", avg: tick_interval - total_used}

    bars =
      systems
      |> render_bars(tick_interval)
      |> add_colors()

    pie_chart = make_pie_chart(systems)

    socket = assign(socket, bars: bars, systems: systems, idle: idle, pie_chart: pie_chart)

    {:ok, socket}
  end

  def render(assigns) do
    ~H"""
    <div>
      <div>
        <svg viewBox="0 0 100 10">
          <%= for %{offset: offset, percent_size: percent_size, color: color} <- @bars do %>
            <.render_bar offset={offset} percent_size={percent_size} color={color} />
          <% end %>
        </svg>
      </div>

      <table>
        <tr>
          <th>Name</th>
          <th>Avg Time</th>
        </tr>
        <%= for %{name: name, avg: avg} <- @systems do %>
          <tr>
            <td><%= get_alias(name) %></td>
            <td><%= avg %> μs</td>
          </tr>
        <% end %>
      </table>

      <%= Contex.Plot.to_svg(@pie_chart) %>
    </div>
    """
  end

  defp get_alias(module) when is_atom(module) do
    module
    |> Module.split()
    |> List.last()
  end

  defp get_alias(other) when is_binary(other), do: other

  defp render_bars(systems, tick_interval, done \\ [], offset \\ 0)

  defp render_bars([], _tick_interval, done, offset) do
    bar = %{offset: offset, percent_size: 100 - offset}

    [bar | done]
  end

  defp render_bars([system | rest], tick_interval, done, offset) do
    %{name: _name, min: _min, max: _max, avg: avg} = system
    percent_size = 100 * avg / tick_interval

    bar = %{offset: offset, percent_size: percent_size}

    render_bars(rest, tick_interval, [bar | done], offset + percent_size)
  end

  defp add_colors(systems) do
    colors = ColorStream.hex(saturation: 0.5, value: 0.95)

    Enum.zip_with(systems, colors, fn system, color ->
      Map.put(system, :color, "##{color}")
    end)
  end

  def render_bar(assigns) do
    ~H"""
    <rect x={@offset} y="0" width={@percent_size} height="100%" fill={@color} />
    """
  end

  defp make_pie_chart(systems) do
    data = Enum.map(systems, &[get_alias(&1.name), &1.avg])

    dataset = Contex.Dataset.new(data, ["Name", "Avg Runtime"])

    opts = [
      mapping: %{category_col: "Name", value_col: "Avg Runtime"},
      legend_setting: :legend_right,
      data_labels: true,
      title: "System Runtimes"
    ]

    Contex.Plot.new(dataset, Contex.PieChart, 600, 400, opts)
  end
end


lib/ecsx/live_dashboard/page.ex
defmodule ECSx.LiveDashboard.Page do
  @moduledoc false
  use Phoenix.LiveDashboard.PageBuilder

  alias ECSx.LiveDashboard.Store
  alias Phoenix.LiveDashboard.PageBuilder

  @base_tabs [:systems, :components]

  @impl PageBuilder
  def mount(params, _session, socket) do
    case params["nav"] do
      nil ->
        to = live_dashboard_path(socket, socket.assigns.page, nav: :systems)
        {:ok, push_navigate(socket, to: to)}

      nav ->
        socket =
          socket
          |> assign_systems()
          |> assign_components()
          |> assign(
            selector: nav,
            tabs: @base_tabs,
            component_type: nil,
            component_table: []
          )

        {:ok, socket}
    end
  end

  defp assign_systems(socket) do
    systems = Store.get_systems()
    tick_interval = div(1_000_000, ECSx.tick_rate())

    total_used =
      systems
      |> Enum.map(& &1.avg)
      |> Enum.sum()

    idle = %{name: "Idle", avg: tick_interval - total_used}

    percent_used = 100 * total_used / tick_interval

    pie_chart = make_pie_chart(systems)

    assign(socket, systems: systems, idle: idle, pie_chart: pie_chart, percent_used: percent_used)
  end

  defp assign_components(socket) do
    manager_module = ECSx.manager()
    component_types_list = manager_module.components()

    component_table_row_counts =
      Map.new(component_types_list, fn type ->
        {type, :ets.info(type, :size)}
      end)

    size =
      component_types_list
      |> Enum.map(&:ets.info(&1, :memory))
      |> Enum.sum()
      |> words_to_bytes()
      |> format_size()

    assign(socket, components: component_table_row_counts, components_total_size: size)
  end

  defp words_to_bytes(words) do
    words * :erlang.system_info(:wordsize)
  end

  defp format_size(bytes) do
    "#{bytes} bytes"
  end

  defp assign_component_table(socket) do
    case socket.assigns.component_type do
      nil -> assign(socket, component_table: [])
      type -> assign(socket, component_table: :ets.tab2list(type))
    end
  end

  @impl PageBuilder
  def menu_link(_, _) do
    {:ok, "ECSx"}
  end

  @impl PageBuilder
  def handle_refresh(socket) do
    case socket.assigns.selector do
      "systems" -> {:noreply, assign_systems(socket)}
      "components" -> {:noreply, assign_components(socket)}
      "component" -> {:noreply, assign_component_table(socket)}
    end
  end

  @impl PageBuilder
  def handle_params(params, _, socket) do
    socket =
      case params["nav"] do
        nil ->
          assign(socket,
            selector: "systems",
            component_type: nil,
            component_table: [],
            tabs: @base_tabs
          )

        "component" ->
          if type = params["type"] do
            component_type =
              socket.assigns.components
              |> Map.keys()
              |> Enum.find(fn full_type -> get_alias(full_type) == type end)

            socket
            |> assign(selector: params["nav"], component_type: component_type)
            |> assign(tabs: @base_tabs ++ [:component])
            |> assign_component_table()
          else
            path =
              live_dashboard_path(socket, socket.assigns.page,
                nav: :component,
                type: get_alias(socket.assigns.component_type)
              )

            push_patch(socket, to: path)
          end

        nav when nav in ["systems", "components"] ->
          assign(socket,
            selector: nav,
            component_type: nil,
            component_table: [],
            tabs: @base_tabs
          )
      end

    {:noreply, socket}
  end

  @impl PageBuilder
  def render(assigns) do
    ~H"""
    <.live_nav_bar id="ecsx_nav_bar" page={@page}>
      <:item :for={tab <- @tabs} name={to_string(tab)} label={format_nav_name(tab)} method="patch">
        <div></div>
      </:item>
    </.live_nav_bar>
    <div style="display: flex; flex-direction: column; align-items: center;">
      <%= case @selector do %>

      <% "systems" -> %>
        <div style="width: 50%; margin-bottom: 4rem;">
          <p style="font: bold; font-size: 24px;">
            Total System Load: <%= percent(@percent_used) %>
          </p>
          <div style="height: 50px; width: 100%; border: solid; border-radius: 10px;">
            <div id="percent-used" style={"background-color: #{loadbar_color(@percent_used)}; height: 100%; width: #{@percent_used}%"} />
          </div>
        </div>

        <div style="width: 90%; display: flex; gap: 10%; margin-bottom: 32px;">
          <div style="width: 45%; font-size: 12px;">
            <%= Contex.Plot.to_svg(@pie_chart) %>
          </div>

          <table style="font-size: 16px; width: 45%;">
            <tr>
              <th>Name</th>
              <th>Avg run time</th>
            </tr>
            <%= for %{name: name, avg: avg} <- @systems do %>
              <tr>
                <td><%= get_alias(name) %></td>
                <td><%= avg %> μs</td>
              </tr>
            <% end %>
          </table>
        </div>

      <% "components" -> %>
        <table style="font-size: 16px; width: 50%;">
          <tr>
            <th>Component type</th>
            <th>Count</th>
          </tr>
          <%= for {name, count} <- @components do %>
            <tr>
              <td><%= get_alias(name) %></td>
              <td>
                <.link patch={live_dashboard_path(@socket, @page, nav: :component, type: get_alias(name))}>
                  <%= count %>
                </.link>
              </td>
            </tr>
          <% end %>
        </table>

      <% "component" -> %>
        <p style="font: bold; font-size: 24px;">
          <%= get_alias(@component_type) %> Components
        </p>
        <table style="font-size: 16px; width: 50%;">
          <tr>
            <th>Entity</th>
            <th>Value</th>
            <th>Persist?</th>
          </tr>
          <%= for {entity, value, persist} <- @component_table do %>
            <tr>
              <td><%= entity %></td>
              <td><%= value %></td>
              <td><%= persist %></td>
            </tr>
          <% end %>
        </table>

      <% end %>
    </div>
    """
  end

  defp format_nav_name(:systems), do: "Systems"
  defp format_nav_name(:components), do: "Components"
  defp format_nav_name(:component), do: "Component"

  defp loadbar_color(percent_used) when percent_used > 50, do: "red"
  defp loadbar_color(percent_used) when percent_used > 30, do: "orange"
  defp loadbar_color(_percent_used), do: "green"

  defp get_alias(module) when is_atom(module) do
    module
    |> Module.split()
    |> List.last()
  end

  defp percent(n) do
    n
    |> round()
    |> Integer.to_string()
    |> then(&[&1, "%"])
  end

  defp make_pie_chart(systems) do
    data = Enum.map(systems, &[get_alias(&1.name), &1.avg])

    dataset = Contex.Dataset.new(data, ["Name", "Avg Runtime"])

    opts = [
      mapping: %{category_col: "Name", value_col: "Avg Runtime"},
      legend_setting: :legend_right,
      data_labels: false,
      title: "System Runtimes"
    ]

    Contex.Plot.new(dataset, Contex.PieChart, 600, 400, opts)
  end
end
